File: web/src/api/goal.ts
import http from './http'
import type { ApiResponse, Goal, GoalPriority, GoalSource, GoalStatus } from '@/types'

export interface CreateGoalPayload {
  title: string
  parent_goal_id?: number
  description?: string
  status?: GoalStatus
  source?: GoalSource
  priority?: GoalPriority
  deadline?: string
  start_date?: string
  target_date?: string
  review_date?: string
  outcome?: string
  success_criteria?: string[]
  motivation?: string
  progress?: number
}

export interface UpdateGoalPayload extends Partial<CreateGoalPayload> {
  status?: Goal['status']
  progress?: number
}

export interface GenerateGoalPayload {
  prompt: string
  context_text?: string
}

function normalizeDate(value?: string) {
  if (!value) return undefined
  if (value.includes('T')) return value
  return new Date(`${value}T00:00:00Z`).toISOString()
}

function normalizeGoalDates<T extends CreateGoalPayload | UpdateGoalPayload>(data: T): T {
  return {
    ...data,
    deadline: normalizeDate(data.deadline),
    start_date: normalizeDate(data.start_date),
    target_date: normalizeDate(data.target_date),
    review_date: normalizeDate(data.review_date),
  }
}

export const goalApi = {
  list: (userID: number) =>
    http.get<ApiResponse<Goal[]>>(`/u/${userID}/goals`).then((r) => r.data.data),

  create: (userID: number, data: CreateGoalPayload) =>
    http.post<ApiResponse<Goal>>(`/u/${userID}/goals`, normalizeGoalDates(data)).then((r) => r.data.data),

  generate: (userID: number, data: GenerateGoalPayload) =>
    http.post<ApiResponse<Goal>>(`/u/${userID}/goals/generate`, data).then((r) => r.data.data),

  getById: (id: number) =>
    http.get<ApiResponse<Goal>>(`/goals/${id}`).then((r) => r.data.data),

  update: (id: number, data: UpdateGoalPayload) =>
    http.put<ApiResponse<Goal>>(`/goals/${id}`, normalizeGoalDates(data)).then((r) => r.data.data),

  delete: (id: number) => http.delete(`/goals/${id}`),
}

File: internal/domain/goal.go
package domain

import (
	"time"

	"gorm.io/gorm"
)

type GoalStatus string
type GoalPriority string
type GoalSource string

const (
	GoalStatusDraft     GoalStatus = "draft"
	GoalStatusActive    GoalStatus = "active"
	GoalStatusCompleted GoalStatus = "completed"
	GoalStatusArchived  GoalStatus = "archived"
	GoalStatusAbandoned GoalStatus = "abandoned"

	GoalPriorityHigh   GoalPriority = "high"
	GoalPriorityMedium GoalPriority = "medium"
	GoalPriorityLow    GoalPriority = "low"

	GoalSourceManual GoalSource = "manual"
	GoalSourceLLM    GoalSource = "llm"
)

// Goal represents a user's desired result and completion contract.
type Goal struct {
	ID        uint           `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`

	UserID       uint         `json:"user_id" gorm:"not null;index"`
	ParentGoalID *uint        `json:"parent_goal_id,omitempty" gorm:"index"`
	Title        string       `json:"title" gorm:"not null"`
	Description  string       `json:"description"`
	Status       GoalStatus   `json:"status" gorm:"default:'active'"`
	Source       GoalSource   `json:"source" gorm:"default:'manual'"`
	Priority     GoalPriority `json:"priority" gorm:"default:'medium'"`
	Deadline     *time.Time   `json:"deadline,omitempty"`
	StartDate    *time.Time   `json:"start_date,omitempty"`
	TargetDate   *time.Time   `json:"target_date,omitempty"`
	ReviewDate   *time.Time   `json:"review_date,omitempty"`
	Outcome      string       `json:"outcome"`
	// SuccessCriteria stores concrete completion standards for the goal.
	SuccessCriteria []string `json:"success_criteria" gorm:"serializer:json"`
	Motivation      string   `json:"motivation"`
	Progress        int      `json:"progress" gorm:"default:0"` // 0-100

	ParentGoal *Goal  `json:"parent_goal,omitempty" gorm:"foreignKey:ParentGoalID"`
	ChildGoals []Goal `json:"child_goals,omitempty" gorm:"foreignKey:ParentGoalID"`
	Tasks      []Task `json:"tasks,omitempty" gorm:"foreignKey:GoalID"`
}
